### PR TITLE
Add support for generating tokens from user-assigned identities

### PR DESCRIPTION
**Pull request by niontive**
1. To generalize the `credentialsObject`, create a new object specific to user-assigned managed identities. This struct will support methods that depend on the `credentialsObject` representing a user-assigned MSI.
2. Add method to new user assigned managed identity object to get token based on resource ID.

### CODE REVIEW

File: pkg/dataplane/identity_test.go [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+			t.Parallel()
+			if _, err := getClientCertificateCredential(tc.identity, tc.cloud); !errors.Is(err, tc.expectedErr) {
+				t.Errorf("expected error: `%s` but got: `%s`", tc.expectedErr, err)
+			}
+		})
+	}
+
+}
+
+func TestValidateUserAssignedMSI(t *testing.T) {
```
Comment (niontive) [inline pkg/dataplane/identity_test.go:228]: This test is copied/pasted from `client_test.go`.

File: pkg/dataplane/identity.go [hunk 2]
```diff
@@ -1,7 +1,133 @@
...
+	}
+	rsaKey, isRsaKey := key.(*rsa.PrivateKey)
+	if !isRsaKey {
+		return nil, errNotRSAKey
+	}
+
+	return azidentity.NewClientCertificateCredential(*identity.TenantID, *identity.ClientID, []*x509.Certificate{crt}, rsaKey, opts)
+}
+
+func validateUserAssignedMSIs(identities []*swagger.NestedCredentialsObject, resourceIDs []string) error {
```
Comment (niontive) [inline pkg/dataplane/identity.go:97]: From `client.go`

File: pkg/dataplane/identity.go [hunk 3]
```diff
@@ -1,7 +1,133 @@
...
+		},
+	}
+	opts.Cloud.ActiveDirectoryAuthorityHost = *identity.AuthenticationEndpoint
+
+	// Parse the certificate and private key from the base64 encoded secret
+	decodedSecret, err := base64.StdEncoding.DecodeString(*identity.ClientSecret)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errDecodeClientSecret, err)
+	}
+	key, crt, err := pkcs12.Decode(decodedSecret, "")
```
Comment (niontive) [inline pkg/dataplane/identity.go:85]: We could use az identity's function: https://github.com/Azure/azure-sdk-for-go/blob/48d39a82091b3aebe3df3505bd5372294b3461ed/sdk/azidentity/client_certificate_credential.go#L94

I don't b/c MSI team says to treat certs like a black box, and that function doesn't support SHA256 MAC so I'd rather have better compat than reusing azidentity.

File: pkg/dataplane/identity.go [hunk 4]
```diff
@@ -1,7 +1,133 @@
...
+	CredentialsObject
+}
+
+// This method may be used by clients to check if they can use the object as a user-assigned managed identity
+// Ex: get credentials object from key vault store and check if it is a user-assigned managed identity to call client for object refresh.
+func (c CredentialsObject) IsUserAssigned() bool {
+	return len(c.ExplicitIdentities) > 0
+}
+
+func (u UserAssignedIdentities) GetCredential(resourceID string) (*azidentity.ClientCertificateCredential, error) {
```
Comment (weinong) [inline pkg/dataplane/identity.go:44]: could use a comment here.

File: internal/test/util.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
...zu9g8SYdz1wSVXFLyiTuxlsQ40FGZWpmea/iVdwzh6j51gMGQd7Npug6P8RO9N8Jy50KTzz9i+UUa9otVgrims7sLPzm7KDXHubQzl0XQu8rUfEYu57kRVT7w6Rp8ZPe6LITaQBihNi1gzpkRgSvq7qKBCS7ZuC3NhXgVelXg0V/24f8Z3qbauofxlRCDNTUZjzpyDO/uGniHhd7qwSruuaM9JpnY7JRQguwTkDyh9G1JOaF9ZhPb81fBtYJt8QEMHIQhwYEN8GeJE0Vnm94j9FLCqBRHdMy579bEDWH14mDVJioMZr6sJlp6yiRutnB0l3ee/rgxVW84qJGtRywMMW3c7znoBggNWDfaX+4w00lSXQRoAQr32HtDIm3q0iZkYgqevDQiaTVPv9yWoU/AMrslvm/LYxvXBens/fQU/NQ4ECD7Xb0n7CSKUbV7JdqsWh/L8q1Cuv795Re6omjgnxD+1+8+vCu/DP3kiHlC0kvLDpOOmP5hxyiMsfs6Tlgtd+HZeBRFsSSvGKdgU5FWe6hP0WVgahecTG0MdJ1HHyJ1ggxvCwKS8IlQ1x9QDqmRt7odH3BuLc3v10/Z+4kPsg7XXsbTT1COPl4SrWKqpWBhfqaeEmVoO4WJdjEql7QEfHgkuUc9xjU3n3BxI+Lipkrjzyuz+vv768tlmqu2H4Tb55bzF+byfDINnRgUCHpwucVaxMRanq8LZZnnGfxOuLwGtn6NYjBLIowazG0pjFA77wWh8DCghInj+rPdLPzhRvWxRp/DW8v24D6fcZBI0GQRmOEwM+Gv1n37RlepYg32KjFF1AJJW0wbZRc2BwuEvf0D5ht1rSQ9VjFd2G78oRBeN5F6zkIy7QsRSwESfmS0LB2ZTTdACBtX3n9zX7kwvsNPICySaWOkaxVl+sJMRmhZYPSSPYFFKLly+D2su5oSIctCQG8qO/GQzJWCdQjqn7/WhftYQdmv+kWqQj3zikRCeSMDduXR/nb5IDWx6UQXkCCSZC0Sjl0fVhWzFQMCMGCSqGSIb3DQEJFTEWBBRMnFjR8YjQYmIJvi83EyaCU1yuujApBgkqhkiG9w0BCRQxHB4aAEEAbABpAGEAcwAgAG8AZgAgAGMAZQByAHQwMTAhMAkGBSsOAwIaBQAEFNbPiz55aOKuhh8J9ZFqV84OoBZwBAhu3BRUgoInHQICCAA="
```
Comment (niontive) [inline internal/test/util.go:7]: This cert is a test one I found in a team internal repo. Not a real secret.

**Review by weinong — COMMENTED** [review]
LGTM

File: internal/test/util.go (hunk 5 above)
Comment (gouthamMN) [reply internal/test/util.go:7]: Is it better to read it from a file (something like config or secret) rather than here?

**Review by rajdeepc2792 — APPROVED** [review]

File: pkg/dataplane/identity.go [hunk 6]
```diff
@@ -1,7 +1,133 @@
...
+	for _, id := range u.ExplicitIdentities {
+		if id != nil && id.ResourceID != nil {
+			if *id.ResourceID == resourceID {
+				return getClientCertificateCredential(*id, u.cloud)
```
Comment (gouthamMN) [inline pkg/dataplane/identity.go:48]: isn't `cloud ` is defined inside `u.CredentialsObject.cloud`

**Review by gouthamMN — APPROVED** [review]
lgtm!

File: pkg/dataplane/client_test.go [hunk 7]
```diff
@@ -51,74 +49,74 @@ func TestGetUserAssignedIdentities(t *testing.T) {
...
 			goMockCall:  func(swaggerClient *mock.MockswaggerMSIClient) {},
-			request:     UserAssignedMSIRequest{ResourceIDs: []string{validResourceID}, TenantID: validTenantID},
+			request:     UserAssignedMSIRequest{ResourceIDs: []string{test.ValidResourceID}, TenantID: test.ValidTenantID},
 			expectedErr: errInvalidRequest,
 		},
 		{
 			name:        "IdenityURL not a URL",
 			goMockCall:  func(swaggerClient *mock.MockswaggerMSIClient) {},
-			request:     UserAssignedMSIRequest{IdentityURL: "bogus", ResourceIDs: []string{validResourceID}, TenantID: validTenantID},
+			request:     UserAssignedMSIRequest{IdentityURL: "bogus", ResourceIDs: []string{test.ValidResourceID}, TenantID: test.ValidTenantID},
```
[the remark this answers is not in the document]
Comment (nwnt) [reply pkg/dataplane/client_test.go:58]: There are other places that use the same hardcoded value.  Those should be changed as well.

File: internal/test/util.go (hunk 5 above)
Comment (nwnt) [reply internal/test/util.go:7]: It's unclear from the name alone whether it's a private or public cert.

File: pkg/dataplane/identity.go (hunk 3 above)
Comment (niontive) [reply pkg/dataplane/identity.go]: Discussed this offline - we'll use azidentity for now and revert this change if we encounter issues during canary testing. Added code comment to highlight this issue.

File: internal/test/util.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
```
Comment (nwnt) [inline internal/test/util.go:9]: Do these constants need to be exposed?

**Review by rajdeepc2792 — APPROVED** [review]

**Review by gouthamMN — APPROVED** [review]
lgtm

**Pull request merged**